lang_java/tree_sitter/tree_sitter_ast_java_grammar_n.ml

```diff
@@ -0,0 +1,695 @@
+
+(* 
+ * Yoann Padioleau, Sharon Lin
+ * 2020 initial draft
+ *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(*
+ * An AST for Java.
+ *
+ *)
+ (*****************************************************************************)
+(* The Tree-sitter AST java related types *)
+(*****************************************************************************)
+(* ------------------------------------------------------------------------- *)
+(* Token/info *)
+(* ------------------------------------------------------------------------- *)
+
+type 'a wrap  = 'a * string
+
+type _literal =
+| Decimal of decimal_integer_literal
+| Hex of hex_integer_literal
+| Octal of octal_integer_literal
+| Binary of binary_integer_literal
+| Decimal1 of decimal_floating_point_literal
+| Hex1 of hex_floating_point_literal
+| Bool of string wrap
+| Character of character_literal
+| String of string_literal
+| Null of null_literal
+
+
+and decimal_integer_literal = string wrap
+
+and hex_integer_literal = string wrap
+
+and octal_integer_literal = string wrap
+
+and binary_integer_literal = string wrap
+
+and decimal_floating_point_literal = string wrap
+
+and hex_floating_point_literal = string wrap
+
+and character_literal = string wrap
+
+and string_literal = string wrap
+
+and null_literal = string wrap
+
+and _expression =
+| Assignment of assignment_expression
+| Binary of binary_expression
+| Instanceof of instanceof_expression
+| Lambda of lambda_expression
+| Ternary of ternary_expression
+| Update of update_expression
+| Ambiguous of _ambiguous_name
+| Primary of _primary
+| Unary of unary_expression
+| Cast of cast_expression
+
+
+and cast_expression = {
+  t_type: _type;
+  t_type1: _type list;
+  value: _expression;
+}
+
+and assignment_expression = {
+  left: _ambiguous_name * field_access * array_access;
+  operator: string wrap;
+  right: _expression;
+}
+
+and binary_expression = string wrap
+
+
+and instanceof_expression = {
+  left: _expression;
+  right: _type;
+}
+
+and lambda_expression = {
+  body: interm15 option;
+  parameters: interm16 option;
+}
+
+and interm15 =
+| Expression of _expression
+| Block of block
+
+and interm16 =
+| Formal of formal_parameters
+| Identifier of identifier
+| Inferred of inferred_parameters
+
+
+and inferred_parameters = identifier * ( identifier list)
+
+and ternary_expression = {
+  condition: _expression;
+  consequence: _expression;
+  alternative: _expression;
+}
+
+and unary_expression = string wrap
```

Comment:
again it should be like for binary_expression
and unary_expression = { operator2: string wrap; operand: _expression }